# CPSC 330 - Applied Machine Learning 

## Homework 8: Time series

### Associated lectures: [Lectures 19 and 20](https://github.com/UBC-CS/cpsc330-2024s/tree/main/lectures)

**See PrairieLearn for _due date_ and _submission_**

## Imports

In [1]:
from hashlib import sha1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import r2_score

## Submission instructions
<hr>

_points: 2_

You will receive marks for correctly submitting this assignment. To submit this assignment, follow the instructions below:

- **You may work on this assignment in a group (group size $\leq$ 4) and submit your assignment as a group.** 
- Below are some instructions on working as a group.  
    - The maximum group size is 4. 
    - You can choose your own group members. 
    - Use group work as an opportunity to collaborate and learn new things from each other. 
    - Be respectful to each other and make sure you understand all the concepts in the assignment well. 
    - It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. [Here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members) are some instructions on adding group members in Gradescope.  
- Be sure to follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2024s/blob/main/docs/homework_instructions.md).
- Upload the .ipynb file to PrairieLearn.

<br><br>

## Exercise 1: time series prediction

In this exercise we'll be looking at a [dataset of avocado prices](https://www.kaggle.com/neuromusic/avocado-prices). You should start by downloading the dataset. We will be forcasting average avocado price for the next week. 

In [2]:
df = pd.read_csv("data/avocado.csv", parse_dates=["Date"], index_col=0)
df.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0 2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1 2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2 2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3 2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4 2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [3]:
df.shape

(18249, 13)

In [4]:
df["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [5]:
df["Date"].max()

Timestamp('2018-03-25 00:00:00')

It looks like the data ranges from the start of 2015 to March 2018 (~2 years ago), for a total of 3.25 years or so. Let's split the data so that we have a 6 months of test data.

In [6]:
split_date = '20170925'
df_train = df[df["Date"] <= split_date]
df_test  = df[df["Date"] >  split_date]

In [7]:
assert len(df_train) + len(df_test) == len(df)

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 How many time series? 

_Points:_ 4

In the [Rain in Australia](https://www.kaggle.com/datasets/jsphyg/weather-dataset-rattle-package) dataset from lecture, we had different measurements for each Location. How about this avocado sales dataset? For which categorical feature(s), if any, do we have separate measurements? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.1
    
</div>

We have Separate Measurements for regions and types
and 54*2 = 108 time series

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18249 entries, 0 to 11
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          18249 non-null  datetime64[ns]
 1   AveragePrice  18249 non-null  float64       
 2   Total Volume  18249 non-null  float64       
 3   4046          18249 non-null  float64       
 4   4225          18249 non-null  float64       
 5   4770          18249 non-null  float64       
 6   Total Bags    18249 non-null  float64       
 7   Small Bags    18249 non-null  float64       
 8   Large Bags    18249 non-null  float64       
 9   XLarge Bags   18249 non-null  float64       
 10  type          18249 non-null  object        
 11  year          18249 non-null  int64         
 12  region        18249 non-null  object        
dtypes: datetime64[ns](1), float64(9), int64(1), object(2)
memory usage: 1.9+ MB


In [9]:
#get unique regions and types (both are categorical)
unique_regions = df['region'].unique()
unique_types = df['type'].unique()

In [10]:
len(unique_regions)

54

In [11]:
len(unique_types)

2

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Equally spaced measurements? 

_Points:_ 4

In the Rain in Australia dataset, the measurements were generally equally spaced but with some exceptions. How about with this dataset? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.2
    
</div>

**The Measurements are generally squally spaced with an interval of 7 days for each region, type combination. Indicating a weekly data measurement (As seen in the Code below)**

In [12]:
sorted_df = df.sort_values(by=['Date','region'])
sorted_df.head()

Date  AveragePrice  Total Volume       4046       4225      4770  \
51 2015-01-04          1.22      40873.28    2819.50   28287.42     49.90   
51 2015-01-04          1.79       1373.95      57.42     153.88      0.00   
51 2015-01-04          1.00     435021.49  364302.39   23821.16     82.15   
51 2015-01-04          1.76       3846.69    1500.15     938.35      0.00   
51 2015-01-04          1.08     788025.06   53987.31  552906.04  39995.03   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53          0.0  conventional  2015   
51     1162.65     1162.65        0.00          0.0       organic  2015   
51    46815.79    16707.15    30108.64          0.0  conventional  2015   
51     1408.19     1071.35      336.84          0.0       organic  2015   
51   141136.68   137146.07     3990.61          0.0  conventional  2015   

                 region  
51               Albany  
51               Albany  
51              Atlanta  
51              Atlanta  
51  BaltimoreWashington

In [13]:
# Code from lecture 19
count = 0 
for name, group in df.groupby(['region', 'type']):
    print("%-30s %s" % (name, group["Date"].sort_values().diff().value_counts()))
    print('\n')
    count+=1
    if count == 5: 
        break

('Albany', 'conventional')     Date
7 days    168
Name: count, dtype: int64


('Albany', 'organic')          Date
7 days    168
Name: count, dtype: int64


('Atlanta', 'conventional')    Date
7 days    168
Name: count, dtype: int64


('Atlanta', 'organic')         Date
7 days    168
Name: count, dtype: int64


('BaltimoreWashington', 'conventional') Date
7 days    168
Name: count, dtype: int64




<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Interpreting regions 

_Points:_ 4

In the Rain in Australia dataset, each location was a different place in Australia. For this dataset, look at the names of the regions. Do you think the regions are also all distinct, or are there overlapping regions? Justify your answer by referencing the data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

**It looks like there is a lot of unique regions (State names), however there is definitely overlap, as there are some regions that specify the general area (e.g. West, SouthCentral, Southeast, TotalUS). These all are not unique regions and WILL overlap with the other regions in our set**

In [14]:
unique_regions

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

<!-- END QUESTION -->

<br><br>

We will use the entire dataset despite any location-based weirdness uncovered in the previous part.

We will be trying to forecast the avocado price. The function below is adapted from [Lecture 19](https://github.com/UBC-CS/cpsc330-2023W2/blob/main/lectures/mathias/19_time-series.ipynb), with some improvements.

In [15]:
def create_lag_feature(df, orig_feature, lag, groupby, new_feature_name=None, clip=False):
    """
    Creates a new feature that's a lagged version of an existing one.
    
    NOTE: assumes df is already sorted by the time columns and has unique indices.
    
    Parameters
    ----------
    df : pandas.core.frame.DataFrame
        The dataset.
    orig_feature : str
        The column name of the feature we're copying
    lag : int
        The lag; negative lag means values from the past, positive lag means values from the future
    groupby : list
        Column(s) to group by in case df contains multiple time series
    new_feature_name : str
        Override the default name of the newly created column
    clip : bool
        If True, remove rows with a NaN values for the new feature
    
    Returns
    -------
    pandas.core.frame.DataFrame
        A new dataframe with the additional column added.
        
    """
        
    if new_feature_name is None:
        if lag < 0:
            new_feature_name = "%s_lag%d" % (orig_feature, -lag)
        else:
            new_feature_name = "%s_ahead%d" % (orig_feature, lag)
    
    new_df = df.assign(**{new_feature_name : np.nan})
    for name, group in new_df.groupby(groupby):        
        if lag < 0: # take values from the past
            new_df.loc[group.index[-lag:],new_feature_name] = group.iloc[:lag][orig_feature].values
        else:       # take values from the future
            new_df.loc[group.index[:-lag], new_feature_name] = group.iloc[lag:][orig_feature].values
            
    if clip:
        new_df = new_df.dropna(subset=[new_feature_name])
        
    return new_df

We first sort our dataframe properly:

In [16]:
df_sort = df.sort_values(by=["region", "type", "Date"]).reset_index(drop=True)
df_sort

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   
18248 2018-03-25          1.62      15303.40  2325.30   2171.66    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   
18248    10806.44    10569.80      236.64          0.0       organic  2018   

                 region  
0                Albany  
1                Albany  
2                Albany  
3                Albany  
4                Albany  
...                 ...  
18244  WestTexNewMexico  
18245  WestTexNewMexico  
18246  WestTexNewMexico  
18247  WestTexNewMexico  
18248  WestTexNewMexico  

[18249 rows x 13 columns]

We then call `create_lag_feature`. This creates a new column in the dataset `AveragePriceNextWeek`, which is the following week's `AveragePrice`. We have set `clip=True` which means it will remove rows where the target would be missing.

In [17]:
df_hastarget = create_lag_feature(df_sort, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)
df_hastarget

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18243 2018-02-18          1.56      17597.12  1892.05   1928.36    0.00   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18243    13776.71    13553.53      223.18          0.0       organic  2018   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   

                 region  AveragePriceNextWeek  
0                Albany                  1.24  
1                Albany                  1.17  
2                Albany                  1.06  
3                Albany                  0.99  
4                Albany                  0.99  
...                 ...                   ...  
18243  WestTexNewMexico                  1.57  
18244  WestTexNewMexico                  1.54  
18245  WestTexNewMexico                  1.56  
18246  WestTexNewMexico                  1.56  
18247  WestTexNewMexico                  1.62  

[18141 rows x 14 columns]

Our goal is to predict `AveragePriceNextWeek`. 

Let's split the data:

In [18]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

<br><br>

### 1.4 `AveragePrice` baseline 

_Points:_ 4

Soon we will want to build some models to forecast the average avocado price a week in advance. Before we start with any ML though, let's try a baseline. Previously we used `DummyClassifier` or `DummyRegressor` as a baseline. This time, we'll do something else as a baseline: we'll assume the price stays the same from this week to next week. So, we'll set our prediction of "AveragePriceNextWeek" exactly equal to "AveragePrice", assuming no change. That is kind of like saying, "If it's raining today then I'm guessing it will be raining tomorrow". This simplistic approach will not get a great score but it's a good starting point for reference. If our model does worse that this, it must not be very good. 

Using this baseline approach, what $R^2$ do you get on the train and test data?

<div class="alert alert-warning">

Solution_1.4
    
</div>

_Type your answer here, replacing this text._

In [19]:
# AveragePriceNextWeek = AveragePrice
train_baseline = df_train['AveragePrice']
train_baseline.head()

0    1.22
1    1.24
2    1.17
3    1.06
4    0.99
Name: AveragePrice, dtype: float64

In [20]:
train_r2 = None
train_r2 = r2_score(df_train['AveragePriceNextWeek'], train_baseline)
train_r2

0.8285800937261841

In [21]:
test_r2 = None

test_baseline = df_test['AveragePrice']

test_r2 = r2_score(df_test['AveragePriceNextWeek'], test_baseline)
test_r2

0.7631780188583048

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Forecasting average avocado price

_Points:_ 10

Now that the baseline is done, let's build some models to forecast the average avocado price a week later. Experiment with a few approachs for encoding the date. Justify the decisions you make. Which approach worked best? Report your test score and briefly discuss your results.

Benchmark: you should be able to achieve $R^2$ of at least 0.79 on the test set. We got to 0.80, but not beyond that. Let us know if you do better!

Note: because we only have 2 splits here, we need to be a bit wary of overfitting on the test set. Try not to test on it a ridiculous number of times. If you are interested in some proper ways of dealing with this, see for example sklearn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html), which is like cross-validation for time series data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

FUNCTION FROM LECTURE 19

In [22]:
from sklearn.pipeline import make_pipeline
from sklearn.compose import make_column_transformer
def preprocess_features(
    train_df,
    test_df,
    numeric_features,
    categorical_features,
    drop_features,
    target
):

    all_features = set(numeric_features + categorical_features + drop_features + target)
    if set(train_df.columns) != all_features:
        print("Missing columns", set(train_df.columns) - all_features)
        print("Extra columns", all_features - set(train_df.columns))
        raise Exception("Columns do not match")

    numeric_transformer = make_pipeline(
        SimpleImputer(strategy="median"), StandardScaler()
    )
    categorical_transformer = make_pipeline(
        SimpleImputer(strategy="constant", fill_value="missing"),
        OneHotEncoder(handle_unknown="ignore", sparse_output=False),
    )

    preprocessor = make_column_transformer(
        (numeric_transformer, numeric_features),
        (categorical_transformer, categorical_features),
        ("drop", drop_features),
    )
    preprocessor.fit(train_df)
    ohe_feature_names = (
        preprocessor.named_transformers_["pipeline-2"]
        .named_steps["onehotencoder"]
        .get_feature_names_out(categorical_features)
        .tolist()
    )
    new_columns = numeric_features + ohe_feature_names

    X_train_enc = pd.DataFrame(
        preprocessor.transform(train_df), index=train_df.index, columns=new_columns
    )
    X_test_enc = pd.DataFrame(
        preprocessor.transform(test_df), index=test_df.index, columns=new_columns
    )

    y_train = train_df["AveragePriceNextWeek"]
    y_test = test_df["AveragePriceNextWeek"]

    return X_train_enc, y_train, X_test_enc, y_test, preprocessor

**Encoding Date as days since first day**

In [23]:
df_train["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [24]:
numeric_features = ['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'year']
categorical_features = ['type','region']
drop_features = ['Date']
target = ['AveragePriceNextWeek']

# Code from lecture 19
train_df = df_train.assign(
    Month=df_train["Date"].apply(lambda x: x.month_name())
)  # x.month_name() to get the actual string
test_df = df_test.assign(Month=df_test["Date"].apply(lambda x: x.month_name()))
train_df[["Date", "Month"]].sort_values(by="Month")

Date      Month
1304  2017-04-30      April
9869  2016-04-17      April
9025  2016-04-24      April
3332  2017-04-30      April
3331  2017-04-23      April
...          ...        ...
11241 2016-09-04  September
11242 2016-09-11  September
11243 2016-09-18  September
11293 2017-09-03  September
18222 2017-09-24  September

[15441 rows x 2 columns]

In [25]:
# Preprocess the data 
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    train_df, test_df, 
    numeric_features, 
    categorical_features + ["Month"], 
    drop_features,
    target
)

In [26]:
X_train_enc.head()

AveragePrice  Total Volume      4046      4225      4770  Total Bags  \
0     -0.432512     -0.234535 -0.229503 -0.222203 -0.214954   -0.232206   
1     -0.383676     -0.234440 -0.230948 -0.219448 -0.214272   -0.233587   
2     -0.554604     -0.233469 -0.231018 -0.219530 -0.214196   -0.229850   
3     -0.823205     -0.233283 -0.230996 -0.218170 -0.213945   -0.230999   
4     -0.994133     -0.225747 -0.230668 -0.196131 -0.213811   -0.232627   

   Small Bags  Large Bags  XLarge Bags      year  ...  Month_December  \
0   -0.229907   -0.223154    -0.172063 -1.147053  ...             0.0   
1   -0.231513   -0.223789    -0.172063 -1.147053  ...             0.0   
2   -0.226469   -0.224325    -0.172063 -1.147053  ...             0.0   
3   -0.228629   -0.222193    -0.172063 -1.147053  ...             0.0   
4   -0.229930   -0.224856    -0.172063 -1.147053  ...             0.0   

   Month_February  Month_January  Month_July  Month_June  Month_March  \
0             0.0            1.0         0.0         0.0          0.0   
1             0.0            1.0         0.0         0.0          0.0   
2             0.0            1.0         0.0         0.0          0.0   
3             0.0            1.0         0.0         0.0          0.0   
4             1.0            0.0         0.0         0.0          0.0   

   Month_May  Month_November  Month_October  Month_September  
0        0.0             0.0            0.0              0.0  
1        0.0             0.0            0.0              0.0  
2        0.0             0.0            0.0              0.0  
3        0.0             0.0            0.0              0.0  
4        0.0             0.0            0.0              0.0  

[5 rows x 78 columns]

In [27]:
# Implementing RIDGE as our first regression model (using generally good alpha since we are looking more at the date encoding)
rm = Ridge(alpha=1.0)
rm.fit(X_train_enc, y_train)
y_train_pred_rm = rm.predict(X_train_enc)

# Getting r2 Scores for train
train_r2_rm = r2_score(y_train, y_train_pred_rm)
print("Train r2 score:", train_r2_rm)

# Getting score for test
y_test_pred_rm = rm.predict(X_test_enc)
test_r2_rm = r2_score(y_test, y_test_pred_rm)
print("Test r2 score:", test_r2_rm)

Train r2 score: 0.8493687878943418
Test r2 score: 0.8001070748052257


In [28]:
# implementing RANDOM FOREST as another model for regression (using generally good n_estimators)
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_enc, y_train);

# Getting train score
y_train_pred_rf = rf.predict(X_train_enc)
train_r2_rf = r2_score(y_train, y_train_pred_rf)
print("Train r2 score:", train_r2_rf)

# Getting test score
y_test_pred_rf = rf.predict(X_test_enc)
test_r2_rf = r2_score(y_test, y_test_pred_rf)
print("Train r2 score:", test_r2_rf)

Train r2 score: 0.97866767648991
Train r2 score: 0.7662632775055288


<br>**One-Hot encoding of seasons**

In [29]:
def get_season(month):
    # remember this is Australia
    WINTER_MONTHS = ["June", "July", "August"] 
    AUTUMN_MONTHS = ["March", "April", "May"]
    SUMMER_MONTHS = ["December", "January", "February"]
    SPRING_MONTHS = ["September", "October", "November"]
    if month in WINTER_MONTHS:
        return "Winter"
    elif month in AUTUMN_MONTHS:
        return "Autumn"
    elif month in SUMMER_MONTHS:
        return "Summer"
    else:
        return "Fall"

In [30]:
train_df = df_train.assign(Season=train_df["Month"].apply(get_season))
test_df = df_test.assign(Season=test_df["Month"].apply(get_season))

In [31]:
train_df = train_df.assign(
    Month=train_df["Date"].apply(lambda x: x.month_name())
)  # x.month_name() to get the actual string
test_df = test_df.assign(Month=test_df["Date"].apply(lambda x: x.month_name()))
train_df[["Date", "Month"]].sort_values(by="Month")

Date      Month
1304  2017-04-30      April
9869  2016-04-17      April
9025  2016-04-24      April
3332  2017-04-30      April
3331  2017-04-23      April
...          ...        ...
11241 2016-09-04  September
11242 2016-09-11  September
11243 2016-09-18  September
11293 2017-09-03  September
18222 2017-09-24  September

[15441 rows x 2 columns]

In [32]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    train_df,
    test_df,
    numeric_features,
    categorical_features + ["Season"],
    drop_features + ["Month"],
    target
)

In [33]:
# Implementing RIDGE as our first regression model
rm.fit(X_train_enc, y_train)
y_train_pred_rm = rm.predict(X_train_enc)

# Getting r2 Scores for train
train_r2_rm = r2_score(y_train, y_train_pred_rm)
print("Train r2 score:", train_r2_rm)

# Getting score for test
y_test_pred_rm = rm.predict(X_test_enc)
test_r2_rm = r2_score(y_test, y_test_pred_rm)
print("Test r2 score:", test_r2_rm)

Train r2 score: 0.8475540634783671
Test r2 score: 0.7935751790455308


In [34]:
# RANDOM FOREST as another model
rf.fit(X_train_enc, y_train);

# Getting train score
y_train_pred_rf = rf.predict(X_train_enc)
train_r2_rf = r2_score(y_train, y_train_pred_rf)
print("Train r2 score:", train_r2_rf)

# Getting test score
y_test_pred_rf = rf.predict(X_test_enc)
test_r2_rf = r2_score(y_test, y_test_pred_rf)
print("Train r2 score:", test_r2_rf)

Train r2 score: 0.9783087888795411
Train r2 score: 0.761617777290249


### Summary of Data Above
- **Encoding Date Methods Used**:
    - Encoding date as days since minimum date
    - Encoding date as one-hot encoded seasons
- **Regression Models Used**:
    - Ridge
    - RandomForest

**SCORES**:
- **Encoding date as days since minimum date**:
    - **Ridge**
        - Train = 0.85
        - Test = 0.80
    - **RF**
        - Train = 0.98
        - Test = 0.77
- **Encoding date as one hot encoded seasons**:
    - **Ridge**
        - Train = 0.85
        - Test = 0.79
    - **RF**
        - Train = 0.98
        - Test = 0.76

**Conclusion**: It seems that the best performing model was the ridge model with date encoded as days since minimum date, and the ridge model appeared to be performing better than the random forest model in both cases.

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: very short answer questions

Each question is worth 2 points.

<!-- BEGIN QUESTION -->

### 2.1 Time series

_Points:_ 4

The following questions pertain to Lecture 19 on time series data:

1. Sometimes a time series has missing time points or, worse, time points that are unequally spaced in general. Give an example of a real world situation where the time series data would have unequally spaced time points.
2. In class we discussed two approaches to using temporal information: encoding the date as one or more features, and creating lagged versions of features. Which of these (one/other/both/neither) two approaches would struggle with unequally spaced time points? Briefly justify your answer.

<div class="alert alert-warning">

Solution_2.1
    
</div>

1. Earthquakes would be unequally spaced as they are very unpredictable and collecting data on them would not be at a regular interval
2. Lagged features rely on consistent intervals between observations, so if we have unequally spaced time points, the approach of using lagged features would be worse

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Survival analysis

_Points:_ 6

The following questions pertain to Lecture 20 on survival analysis. We'll consider the use case of customer churn analysis.

1. What is the problem with simply labeling customers are "churned" or "not churned" and using standard supervised learning techniques?
2. Consider customer A who just joined last week vs. customer B who has been with the service for a year. Who do you expect will leave the service first: probably customer A, probably customer B, or we don't have enough information to answer?
3. If a customer's survival function is almost flat during a certain period, how do we interpret that?

<div class="alert alert-warning">

Solution_2.2
    
</div>

1. By labeling churned or not churned only using standard techniques, we are ignoring the times of when they churned. So data will be lost about the length of staying, and survival probability.
2. Not enough info to answer (we can't just predict solely on how long they have been with the service)
3. It means that there is a high probability of the customer staying/low churn chance

<!-- END QUESTION -->

<br><br>

## Submission instructions 

**PLEASE READ:** When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from “1” will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using PrairieLearn.
4. Make sure that the plots and output are rendered properly in your submitted file.